### PR TITLE
 Add import from cohort/concept set

### LINKED ISSUE #531: Add import from cohort/conceptset
### Expected behavior

When inside a conceptset in a cohort definition, you should be able to import a concept/conceptset (akin to function on vocabulary tab).

### Actual behavior

v2.3 prevents import of concepts into an open conceptset in a cohort

### PR DESCRIPTION
**Pull request by wivern**
closes #531

### CODE REVIEW

File: js/components/cohort-definition-manager.html [hunk 1]
```diff
@@ -56,6 +56,7 @@
 			<div role="tabpanel" data-bind="css: { active: $component.tabMode() == 'conceptsets' }, eventListener: [{event: 'click', selector: '.repositoryConceptSetItem', callback: onConceptSetTabRespositoryConceptSetSelected}]" class="tab-pane">
 				<div style="padding-bottom: 10px;" class="pull-right">
 					<button type="button" class="btn btn-sm btn-success pull-right" data-bind="click:function() { exportConceptSetsCSV(); }, css: {'disabled': dirtyFlag().isDirty, 'btn-success': !dirtyFlag().isDirty()}"><i class="fa fa-download" aria-hidden="true"></i> Export All Concept Sets To CSV</button>
+					<button type="button" style="margin-right: 3px;" class="btn btn-sm btn-primary pull-right" data-bind="click:function(){ $component.importConceptSet() }, enable: canEdit"><i class="fa fa-upload" aria-hidden="true" /> Import</button>
```
Comment (pavgra) [inline js/components/cohort-definition-manager.html]: no inline styles pls

File: js/components/cohort-definition-manager.html [hunk 2]
```diff
@@ -80,6 +81,13 @@
 											}
 									 }">
 				</table>
+				<div class="paddedWrapper form-group" data-bind="if:isImportingConceptSet()">
+					<div class="heading">Concept Set Expression JSON (<a class="linkish" data-bind="click: function() { $('#textImportConceptSet').val('');}">click to clear</a>)</div>
```
Comment (pavgra) [inline js/components/cohort-definition-manager.html]: Do not inline this pls

$('#textImportConceptSet').val('')

File: js/components/cohort-definition-manager.html [hunk 3]
```diff
@@ -80,6 +81,13 @@
...
+					<textarea class="form-control" rows="5" id="textImportConceptSet"></textarea>
+					<div type="button" data-bind="css: {disabled: model.currentConceptSet()!=null}, click:function() { importConceptSetExpression($('#textImportConceptSet').val()); }" class="btn btn-sm btn-primary pull-right "><i class="fa fa-upload"></i> Import Concept Set Expression</div>
+					<button class="btn btn-sm btn-primary pull-right" style="margin-right: 3px" data-bind="click: function(){ $component.closeConceptSetImport() }">Close Import</button>
```
Comment (pavgra) [inline js/components/cohort-definition-manager.html]: Look above

File: js/components/cohort-definition-manager.html [hunk 4]
```diff
@@ -80,6 +81,13 @@
 											}
 									 }">
 				</table>
+				<div class="paddedWrapper form-group" data-bind="if:isImportingConceptSet()">
+					<div class="heading">Concept Set Expression JSON (<a class="linkish" data-bind="click: function() { $('#textImportConceptSet').val('');}">click to clear</a>)</div>
+					<textarea class="form-control" rows="5" id="textImportConceptSet"></textarea>
+					<div type="button" data-bind="css: {disabled: model.currentConceptSet()!=null}, click:function() { importConceptSetExpression($('#textImportConceptSet').val()); }" class="btn btn-sm btn-primary pull-right "><i class="fa fa-upload"></i> Import Concept Set Expression</div>
```
Comment (pavgra) [inline js/components/cohort-definition-manager.html]: Look above

File: js/components/cohort-definition-manager.html [hunk 5]
```diff
@@ -109,7 +120,7 @@
 						<div class="clear"></div>
 
 						<button type="button" class="btn btn-sm btn-primary pull-right" data-bind="click: function() { $component.closeConceptSet()};">Close Concept Set</button>
-						<button type="button" class="btn btn-sm btn-danger pull-right" data-bind="click:function() { $component.deleteConceptSet()};">Delete Concept Set</button>
+						<button type="button" style="margin-right: 3px;" class="btn btn-sm btn-danger pull-right" data-bind="click:function() { $component.deleteConceptSet()};">Delete Concept Set</button>
```
Comment (pavgra) [inline js/components/cohort-definition-manager.html]: Look above

File: js/components/cohort-definition-manager.html [hunk 6]
```diff
@@ -137,6 +148,30 @@
 					</div>
 					<pre id="includedConceptIdentifierList" data-bind="html: model.currentIncludedConceptIdentifierList, selectOnFocus: {events: ['click']}"></pre>
 				</div>
+
+				<div id="wrapperConceptSetImport" data-bind="if:$component.conceptSetTabMode() == 'import'">
+					<ul class="nav nav-pills">
+						<li role="presentation" data-bind="css: {active: $component.importTabMode() == 'identifiers'}, click: function(){ $component.importTabMode('identifiers'); }">
```
Comment (pavgra) [inline js/components/cohort-definition-manager.html]: 'identifiers' to constants

File: js/components/cohort-definition-manager.html [hunk 7]
```diff
@@ -137,6 +148,30 @@
...
+							<a>Concept Identifiers</a>
+						</li>
+						<li role="presentation" data-bind="css: {active: $component.importTabMode() == 'sourcecodes'}, click: function(){ $component.importTabMode('sourcecodes'); }">
```
Comment (pavgra) [inline js/components/cohort-definition-manager.html]: 'sourcecodes' to constants

File: js/components/cohort-definition-manager.js [hunk 8]
```diff
@@ -562,17 +569,91 @@ define(['knockout', 'text!./cohort-definition-manager.html',
...
+    function appendConcepts(data){
+      data.forEach(function(item){
+        sharedState.selectedConceptsIndex[item.CONCEPT_ID] = 1;
+        sharedState.selectedConcepts.push(self.model.createConceptSetItem(item));
+      });
+      if (self.model.currentCohortDefinition() && self.model.currentConceptSetSource() == "cohort") {
+        var conceptSet = self.model.currentCohortDefinition()
+          .expression()
+          .ConceptSets()
+          .filter(function (item) {
```
Comment (pavgra) [inline js/components/cohort-definition-manager.js]: find

File: js/components/cohort-definition-manager.js [hunk 9]
```diff
@@ -562,17 +569,91 @@ define(['knockout', 'text!./cohort-definition-manager.html',
 			self.closeConceptSet();
 		}
 
+		function createConceptSet() {
```
Comment (pavgra) [inline js/components/cohort-definition-manager.js]: Cannot we move this code to smth like ConceptSetService and make methods more abstract?

Comment (wivern) [reply js/components/cohort-definition-manager.js]: I promise to think about that but let do such refactoring outside from current issue

File: js/components/cohort-definition-manager.html [hunk 10]
```diff
@@ -108,8 +119,8 @@
 
 						<div class="clear"></div>
 
-						<button type="button" class="btn btn-sm btn-primary pull-right" data-bind="click: function() { $component.closeConceptSet()};">Close Concept Set</button>
-						<button type="button" class="btn btn-sm btn-danger pull-right" data-bind="click:function() { $component.deleteConceptSet()};">Delete Concept Set</button>
+						<button type="button" class="btn btn-sm btn-primary pull-right last" data-bind="click: function() { $component.closeConceptSet()};">Close Concept Set</button>
```
Comment (pavgra) [inline js/components/cohort-definition-manager.html]: Resolve this in same way as we did with :last-child

File: js/components/cohort-definition-manager.html [hunk 11]
```diff
@@ -137,6 +148,30 @@
...
+				<div id="wrapperConceptSetImport" data-bind="if:$component.conceptSetTabMode() == 'import'">
+					<ul class="nav nav-pills">
+						<li role="presentation" data-bind="css: {active: $component.importTabMode() == importTabModes.identifiers}, click: function(){ $component.importTabMode('identifiers'); }">
+							<a>Concept Identifiers</a>
+						</li>
+						<li role="presentation" data-bind="css: {active: $component.importTabMode() == importTabModes.sourcecodes}, click: function(){ $component.importTabMode('sourcecodes'); }">
+							<a>Source Codes</a>
+						</li>
+					</ul>
+					<div id="wrapperImportConceptIdentifiers" class="paddedWrapper form-group" data-bind="if:$component.importTabMode() == 'identifiers'">
```
Comment (pavgra) [inline js/components/cohort-definition-manager.html]: let's switch duplicated strings to constants

File: js/components/cohort-definition-manager.html [hunk 12]
```diff
@@ -137,6 +148,30 @@
...
+						<textarea class="form-control" rows="5" id="textImportConceptIdentifiers" data-bind="if:model.currentImportMode() == 'conceptset'"></textarea>
+						<a data-bind="click:function() {importConceptIdentifiers($('#textImportConceptIdentifiers').val().match(/[0-9]+/g));}" class="btn btn-sm btn-primary pull-right"><span class="fa fa-upload"></span> Import Concept Identifiers</a>
+						<div class="clear"></div>
+					</div>
+					<div id="wrapperImportSourcecodes" class="paddedWrapper form-group" data-bind="if:$component.importTabMode() == 'sourcecodes'">
+						<div class="heading">Enter Source Codes (<a class="linkish" data-bind="click: function() { clearImportedConceptSet('#textImportSourcecodes');}">click to clear</a>)</div>
+						<textarea class="form-control" rows="5" id="textImportSourceCodes"></textarea>
+						<a data-bind="click:function() {importSourceCodes($('#textImportSourceCodes').val().match(/[0-9a-zA-Z\.-]+/g));}" class="btn btn-sm btn-primary pull-right"><span class="fa fa-upload"></span> Import Source Codes</a>
```
Comment (pavgra) [inline js/components/cohort-definition-manager.html]: Cannot we get value from #textImportSourceCodes via 2-way data-binding?

File: js/styles/atlas.css [hunk 13]
```diff
@@ -2164,4 +2164,16 @@ div#wrapper_ohdsi a {
 
 .command-pane {
   display: flex;
+}
+
+.cohort-conceptset-button-pane {
```
Comment (pavgra) [inline js/styles/atlas.css]: Could you move this to component specific css file?

File: js/components/cohort-definitions/cohort-definition-manager.html [hunk 14]
```diff
@@ -13,7 +13,7 @@
 			<div class="input-group">
 				<input class="form-control" type="text" data-bind="enabled: canEdit(), textInput: model.currentCohortDefinition().name, css: { emptyInput: !(model.currentCohortDefinition().name() && (model.currentCohortDefinition().name().length > 0)) }"></input>
 				<div class="input-group-btn">
-					<button class="btn btn-success" title="Save cohort definition" title="Save" data-bind="click: save, enable: (dirtyFlag().isDirty() && !isRunning() && canEdit())"><i class="fa fa-save"></i></button>
```
Comment (anthonysena) [inline js/components/cohort-definitions/cohort-definition-manager.html]: Why were the checks (dirtyFlag().isDirty() && !isRunning() && canEdit()) removed?

**Comment by pavgra** [comment]
@anthonysena, I have a feeling, that you've put some comments, but I cannot find any now

**Comment by anthonysena** [comment]
Looking good here - I made a small fix for the promises since we can use the .then() to handle both the success and failure. In terms of the user experience, I think we need to make a change to make this more consistent. For example, here is how the concept import looks on the vocabulary screen:

vocab-import

You have the 3 pill tabs that allow you to import the concept identifiers, source cods and concept set expression. However, on the cohort screen, it is a little different:

cs-import

I'd prefer that instead of having the "Import" button next to the "Create concept set" button, we instead follow the UI from the vocabulary screen and make that another pill tab that allows you to import the concept set. This way there is consistency in the way a person can import either through the vocabulary section or in the cohort definition editor.

And thanks for putting in some time and effort to further clean up the components folder - I like having the cohort definition components in their own folder. I'm looking forward to when we can tackle #611 per @chrisknoll's suggestion

**Comment by pavgra** [comment]
@anthonysena, Concept Identifiers and Source code import works for specific Concept Set. It is concept set related function, which is meaningful only when we can visually associate it with the concept set. So why it cannot be located next to the concept set's Export button?

The same question is about Import Concept Set button. As or me, it is logical, that I go to Concept Sets tab and there have several options (buttons) presenting me a way to add entity into the section.

**Comment by anthonysena** [comment]
Hey @pavgra - what I am aiming for is consistency between the way you can create a concept set via the vocabulary screen or via the cohort definition screen. I'm not opposed to having the "Import" button where you placed it. In this current implementation, clicking that button allows you to import a concept set expression...but what about importing a list of concept IDs or source codes?

Perhaps an alternative approach: leave the Import button at the top - it creates a new concept set and immediately displays the "Import" pill tab which then has the 3 options for import available. Let me know what you think of that approach.

**Pull request merged**